proto/private/rules/proto_toolchain.bzl

```diff
@@ -0,0 +1,57 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A toolchain for Protobuf rules."""
+
+load("//proto/private:native.bzl", proto_common = "native_proto_common")
+
+def _proto_toolchain_impl(ctx):
+    return [
+        platform_common.ToolchainInfo(
+            protoc = ctx.attr._protoc[DefaultInfo].files_to_run,
+            protoc_options = getattr(ctx.fragments.proto, "protoc_options", []),
+        ),
+    ]
+
+def _protoc():
+    if "protoc" in getattr(proto_common, "available_configuration_fields", []):
```

Comment:
I don't like this mechanism, but I also don't know any alternatives. Do you have any idea?

There also an option to use a function here, but I think it can only have other attribute values as parameters and not configuration values.

Reply:
Let me retract previous comment.

The order of flags and toolchains is weird. With this implementation the order is: what's on toolchains attribute, then flag (if available) then default value.

Ideally we don't have a flag and only whats on the toolchain matters.

If we really need the flag it should have precedence over what is in the rule. This means to move it into the rule implementation. And then you also don't have need to expose available_configuration_fields.

Reply:
Right, but you can't just use a label from a flag.

I'd really go for a no-flag option. If somebody wants to use a different protoc, let them define proto_toolchain and then use it with `--extra_toolchains=` flag.

Reply:
Removed the dependency on the fragment fields